WalletWasabi/WabiSabi/Models/MultipartyTransaction/State.cs

```diff
@@ -0,0 +1,8 @@
+namespace WalletWasabi.WabiSabi.Models.MultipartyTransaction
+{
+	public abstract record State
+	{
+		public Construction AssertConstruction() => (Construction)this;
```

Comment:
This is smelly. Why not just having an Assert method that is implemented in the child classes with return this.

Reply:
see #5531

`State Round.CoinjoinState` needs to be of type `Construction` input -> output registration, and then `Finalize` converts it to `Signing`. When the Arena needs to update the state, it asserts that it's of the correct type.

If this is implemented as a virtual method on the base class I don't think it's possible to make the return signature a subclass, C# 9 introduced covariant returns so this should now be allowed but from the caller that's still defined dynamically, the point of these is to have a static type.

In #5531 I'm now refactoring `Round` to do e.g.`AddAlice` which will update the state, and I will probably remove these methods, whose main purpose is to remove that boilerplate since it was spread out, but in `Round` that logic will all be in the same place so I think I will just return to e.g. `(Construction)CoinjoinState` in there.

Reply:
Yes, I see it know. This type knows its children and that's a problem.

What if `State` is an `interface` (renamed as `IState`) with a the common properties `EstimatedVsize`, `EstimatedInputsVsize` and `OutputsVsize`, or simple without any property/method and then you create two extension methods `AssertConstruction` and `AssertSigning` for that `IState`? Something like this:

```c#
public static Construction AssertConstruction(this IState state) =>
   state switch {
       Construction => state,
       _ => throw new InvalidStateException($"{nameof(Construction) state was expected but {state.GetType().Name} was whatever... }");
   };
```

Reply:
Maybe interface is what you are looking for here. Casting this to construction if that is a Signing will produce an exception.

Reply:
I am OK with the Concept, readability is better with Lucas's solution or something similar.

Reply:
ack, I like Lucas's solution better too, adding now.

I hope this is temporary and can be removed eventually, see the response to the fluent builder suggestion

Reply:
hmm, is it really worth it to add another exception class instead of just casting? this is not a user or client facing error, but an internal/[programmer error](https://www.joyent.com/node-js/production/design/errors) so I think the built in runtime error from this kind of cast should be enough

```
static class StateExtensions
{
    public static ConstructionState AssertConstruction(this IState state) => (ConstructionState)state;
    public static SigningState AssertSigning(IState state) => (SigningState)state;
}
```

Or alternatively,

```
public static ConstructionState Assert(IState state) => (ConstructionState)state;
```

in `ConstructionState` and then just use `ConstructionState.Assert(x)`?

Reply:
i went with the last idea since it seems simplest to me, if extension methods are cleaner for some reason let me know and I will amend